### PR TITLE
chore(codecs): Implement encoding support structures

### PR DESCRIPTION
**Pull request by pablosichert**
Part of #8617, #10254 and towards the implementation of RFC #9864.

### CODE REVIEW

File: src/codecs/format/bytes.rs [hunk 1]
```diff
@@ -47,13 +51,65 @@ impl Deserializer for BytesDeserializer {
...
+}
+
+#[typetag::serde(name = "bytes")]
+impl SerializerConfig for BytesSerializerConfig {
+    fn build(&self) -> crate::Result<BoxedSerializer> {
+        Ok(Box::new(BytesSerializer))
+    }
+}
+
+/// Serializer that converts bytes to an `Event`.
```
Comment (lukesteensen) [inline src/codecs/format/bytes.rs:72]:
```suggestion
/// Serializer that converts an `Event` to bytes.
```

File: src/codecs/format/bytes.rs [hunk 2]
```diff
@@ -47,13 +51,65 @@ impl Deserializer for BytesDeserializer {
...
+///
+/// This serializer can be considered as the no-op action for input where no
+/// further encoding has been specified.
```
Comment (lukesteensen) [inline src/codecs/format/bytes.rs:75]: We've caused a decent amount of confusion in the past by considering the message field as a "no-op" or default option for encoding, so I'd rather not replicate that here. Two points on this:

1. I'm not sure it makes sense to have a "no-op" encoder. There fundamentally has to be some operation to turn an event into bytes.
2. A serializer that just takes the message field as-is should be named in way that makes that clear.

Comment (pablosichert) [reply src/codecs/format/bytes.rs]: It refers to that behavior more explicitly now:

https://github.com/vectordotdev/vector/blob/04c368e8fe8e06238c1cdc9ef1e7c8836029a1b6/src/codecs/format/text.rs#L28-L30

File: src/codecs/encoding.rs [hunk 3]
```diff
@@ -1,2 +1,201 @@
...
+}
+
+/// A `Box` containing a `FramingError`.
+pub type BoxedFramingError = Box<dyn FramingError>;
+
+/// Wrap bytes into a frame.
+pub trait Framer:
+    tokio_util::codec::Encoder<Bytes, Error = BoxedFramingError> + DynClone + Debug + Send + Sync
+{
+}
```
Comment (spencergilbert) [inline src/codecs/encoding.rs:63]: Just noting how weird this formatting looks 😆

**Review by spencergilbert — APPROVED** [review]
I'd love to get a better name than `text` here, but don't know if we've found one yet. Sort of feels like a nice place to introduce that change. This otherwise _seems_ good to me.

File: src/codecs/encoding.rs [hunk 4]
```diff
@@ -1,2 +1,201 @@
...
+                Error::SerializingError(error)
+            })?;
+
+        // Frame the serialized event.
+        self.framer
+            .encode(payload.freeze(), buffer)
+            .map_err(|error| {
+                emit!(&EncoderFramingFailed { error: &error });
+                Error::FramingError(error)
+            })
```
Comment (lukesteensen) [inline src/codecs/encoding.rs:171]: I'm wondering about the performance implications here. As is, we'll be guaranteeing one allocation and one copy per event. That doesn't seem like a blocker necessarily, but I don't want us to end up in a position where we need to redesign things in order to make them more efficient. Do you think there's a path with this design to enabling relatively "direct" serialization (i.e. writing straight to the eventual output, without intermediary writes and copies)?

**Review by lukesteensen — COMMENTED** [review]
Overall this seems reasonable, just one question about performance implications.

File: src/codecs/encoding.rs (hunk 4 above)
Comment (pablosichert) [reply src/codecs/encoding.rs:171]: Good point, I think writing to the `buffer` input argument directly should work.

I'll have to examine more close what that means for the error case, though. I'm not sure what the caller currently expects from the provided buffer when encoding fails.

Comment (pablosichert) [reply src/codecs/encoding.rs]: Implemented in d796987980d4908aba287776915ba44b7f73d20e.

The `Framer` now uses `tokio_util::codec::Encoder<()>` to modify the frame in-place.

Also noteworthy that `FramedWrite` does not always pass an empty buffer to `encode`. I handled that case specifically in 9aa870b01ad308ffe32d81d2823781ac0fda1f89 by making sure the `Framers`/`Serializers` get a view to the buffer that is intended for the payload only.

**Comment by pablosichert** [comment]
CI soak tests seems to ignore my label 😢

**Comment by pablosichert** [comment]
I'll just wait for https://github.com/vectordotdev/vector/pull/10730 to land and rebase it on top. Not having much luck hitting the soak lottery right now 😄

**Comment by lukesteensen** [comment]
@spencergilbert 

> I'd love to get a better name than `text` here, but don't know if we've found one yet.

The best I've come up with is something like `raw`/`raw_message` or similar. But I don't think those are particularly great either.

**Comment by spencergilbert** [comment]
> @spencergilbert
> 
> > I'd love to get a better name than `text` here, but don't know if we've found one yet.
> 
> The best I've come up with is something like `raw`/`raw_message` or similar. But I don't think those are particularly great either.

`raw_message` is probably the best I've seen so far 🤷

File: src/codecs/framing/character_delimited.rs [hunk 5]
```diff
@@ -137,17 +135,47 @@ impl Decoder for CharacterDelimitedDecoder {
...
+        Self { delimiter }
+    }
+}
+
+impl Encoder<()> for CharacterDelimitedEncoder {
+    type Error = encoding::BoxedFramingError;
+
+    fn encode(&mut self, _: (), buf: &mut BytesMut) -> Result<(), encoding::BoxedFramingError> {
+        buf.reserve(buf.len() + 1);
+        buf.put_u8(self.delimiter);
```
Comment (lukesteensen) [inline src/codecs/framing/character_delimited.rs:178]: I believe `reserve` reserves the passed number of _additional_ bytes, not total. So you shouldn't need to add `buf.len()` here.

Comment (pablosichert) [reply src/codecs/framing/character_delimited.rs:178]: Oh wow, thanks for catching that. TIL that the same is true for `std` data structures. How did I miss this fact until now 😮

Comment (pablosichert) [reply src/codecs/framing/character_delimited.rs]: That `reserve` in particular seems unnecessary either way

Comment (pablosichert) [reply src/codecs/framing/character_delimited.rs]: Probably got that from C++: https://www.cplusplus.com/reference/vector/vector/reserve/, not looking back 😄

**Comment by pablosichert** [comment]
@lukesteensen @spencergilbert I like that reasoning, went with `raw_message::RawMessageSerializer`.

**Review by lukesteensen — APPROVED** [review]

**Pull request merged**